packages/@react-aria/button/src/useButton.ts

```diff
@@ -33,20 +26,19 @@ export interface ButtonAria<T> {
   isPressed: boolean
 }
 
-// Order with overrides is important: 'button' should be default
-export function useButton(props: AriaButtonProps<'button'>, ref: RefObject<HTMLButtonElement>): ButtonAria<ButtonHTMLAttributes<HTMLButtonElement>>;
-export function useButton(props: AriaButtonProps<'a'>, ref: RefObject<HTMLAnchorElement>): ButtonAria<AnchorHTMLAttributes<HTMLAnchorElement>>;
-export function useButton(props: AriaButtonProps<'div'>, ref: RefObject<HTMLDivElement>): ButtonAria<HTMLAttributes<HTMLDivElement>>;
-export function useButton(props: AriaButtonProps<'input'>, ref: RefObject<HTMLInputElement>): ButtonAria<InputHTMLAttributes<HTMLInputElement>>;
-export function useButton(props: AriaButtonProps<'span'>, ref: RefObject<HTMLSpanElement>): ButtonAria<HTMLAttributes<HTMLSpanElement>>;
-export function useButton(props: AriaButtonProps<ElementType>, ref: RefObject<Element>): ButtonAria<DOMAttributes>;
+export type AriaButtonPropsElementType = 'button' | 'a' | 'div' | 'input' | 'span' | ElementType;
+
+export type HTMLAttributesElementType = HTMLButtonElement | HTMLAnchorElement | HTMLDivElement | HTMLInputElement | HTMLSpanElement;
+
+export type RefObjectElementType = HTMLAttributesElementType | Element;
+
 /**
  * Provides the behavior and accessibility implementation for a button component. Handles mouse, keyboard, and touch interactions,
  * focus behavior, and ARIA props for both native button elements and custom element types.
  * @param props - Props to be applied to the button.
  * @param ref - A ref to a DOM element for the button.
  */
-export function useButton(props: AriaButtonProps<ElementType>, ref: RefObject<any>): ButtonAria<HTMLAttributes<any>> {
+export function useButton<T extends AriaButtonPropsElementType = 'button', V extends RefObjectElementType = HTMLButtonElement>(props: AriaButtonProps<T>, ref: RefObject<V>): ButtonAria<HTMLAttributes<any>> {
```

Comment:
Couple of things I noticed, lemme know if you ran into any problems when trying to resolve them previously:
1. you have `T` and `V` as generics here, which can result in the following if the user provides conflicting types (note the "div" vs "anchor") :
<img width="786" alt="image" src="https://user-images.githubusercontent.com/9661127/203392222-86e010ba-6add-46eb-ac3c-6f266699abc1.png">
 Have you tried just having the generic `T` and then using that to get the type for the RefObject? [useTextField](https://github.com/adobe/react-spectrum/blob/519844201d1966268bc8cf380d6fa07a7666f41e/packages/%40react-aria/textfield/src/useTextField.ts#L86) determines its ref type in this fashion.

2. The `ButtonAria` return type is `HTMLAttributes<any>`, resulting in `buttonProps` being `HTMLAttributes<any>`. Previously with overrides that `ButtonAria` type would match the HTMLAttributes of the element type provided to `useButton`. You may be able to mimic what  does [here](https://github.com/adobe/react-spectrum/blob/519844201d1966268bc8cf380d6fa07a7666f41e/packages/%40react-aria/textfield/src/useTextField.ts#L69) + [here](https://github.com/adobe/react-spectrum/blob/519844201d1966268bc8cf380d6fa07a7666f41e/packages/%40react-aria/textfield/src/useTextField.ts#L31-L33) to get the specific return types again